contrib/endpoints/src/api_manager/firebase_rules/firebase_request.h

```diff
@@ -0,0 +1,116 @@
+
+using ::google::api_manager::utils::Status;
+using TestRulesetResponse = ::google::api_manager::proto::TestRulesetResponse;
+using FunctionCall = TestRulesetResponse::TestResult::FunctionCall;
+using ::google::protobuf::RepeatedPtrField;
+
+namespace google {
+namespace api_manager {
+namespace firebase_rules {
+
+struct HttpRequest {
+  std::string url;
+  std::string method;
+  std::string body;
+  auth::ServiceAccountToken::JWT_TOKEN_TYPE token_type;
+};
+
+// A FirebaseRequest object understands the various http requests that need
+// to be generated as a part of the TestRuleset request and response cycle.
+// Here is the intented use of this code:
+// FirebaseRequest request(...);
+// while(!request.IsDone()) {
+//  std::string url, method, body;
+//
+//  /* The following is not a valid C++ statement. But written so the reader can
+//  get a general idea ... */
+//
+//  (url, method, body, token_type) = request.GetHttpRequest();
+//  std::string body = InvokeHttpRequest(url, method, body,
+//                                       GetToken(token_type));
+//  updateResponse(body);
+// }
+//
+// if (request.RequestStatus.ok()) {
+//  .... ALLOW .....
+// } else {
+//  .... DENY .....
+// }
+class FirebaseRequest {
+ public:
+  // Constructor.
+  FirebaseRequest(const std::string &ruleset_name, ApiManagerEnvInterface *env,
+                  std::shared_ptr<context::RequestContext> context);
+
+  // If the firebase Request calling can be terminated.
+  bool IsDone();
+
+  // Get the request status. This request status is only valid if IsDone is
+  // true.
+  Status RequestStatus();
+
+  // This call should be invoked to get the next http request to execute.
+  HttpRequest GetHttpRequest();
+
+  // The response for previous HttpRequest.
+  void UpdateResponse(const std::string &body);
+
+ private:
+  Status UpdateRulesetRequestBody(
+      const RepeatedPtrField<FunctionCall> &func_calls);
+  Status ProcessTestRulesetResponse(const std::string &body);
+  Status ProcessFunctionCallResponse(const std::string &body);
+  Status CheckFuncCallArgs(const FunctionCall &func);
+  Status AddFunctionMock(proto::TestRulesetRequest *request,
+                         const FunctionCall &func_call);
+  void SetStatus(const Status &status);
+  Status SetNextRequest();
+  bool AllFunctionCallsProcessed();
+  std::vector<std::pair<FunctionCall, std::string>>::const_iterator Find(
+      const FunctionCall &func_call);
+
+  ApiManagerEnvInterface *env_;
+  std::shared_ptr<context::RequestContext> context_;
+  std::string ruleset_name_;
+  std::string service_name_;
+  std::string firebase_server_;
+  Status current_status_;
+  bool is_done_;
+  std::vector<std::pair<FunctionCall, std::string>> funcs_with_result_;
+  RepeatedPtrField<FunctionCall>::const_iterator func_call_iter_;
+  TestRulesetResponse response_;
+  HttpRequest *next_request_;
+  HttpRequest firebase_http_request_;
+  HttpRequest external_http_request_;
```

Comment:
what is the different between external_http_request and firebase_http_request?  can we share them?  Are they run parallel?

Need more comment.